widgetastic_manageiq/__init__.py

```diff
@@ -94,6 +95,83 @@ def currently_selected(self):
 
 
+class SectionedBootstrapSelect(BootstrapSelect):
+    """ Refactor to allow for bootstrap select that have sections. """
+
+    SectionOption = namedtuple("SectionOption", ["text", "value", "section_id"])
+    SECTIONS = ".//li[contains(@class, 'dropdown-header')]"
+    OPTIONS = ".//li[contains(@class, 'data-original-index') and contains(@class, 'data-optgroup')]"
+    SECTION_ELEMENTS = './/li[@data-optgroup="{}"]'
+
+    @property
+    def all_sections(self):
+        return [
+            self.Option(self.browser.text(el), el.get_attribute("data-optgroup"))
+            for el in self.browser.elements(self.SECTIONS, parent=self)
+        ]
+
+    @property
+    def all_options(self):
+        return [
+            self.SectionOption(
+                self.browser.text(
+                    self.browser.element(".//span[contains(@class, 'text')]", parent=el)
+                ),
+                el.get_attribute("data-original-index"),
+                el.get_attribute("data-optgroup"),
+            )
+            for el in self.browser.elements("./div/ul/li", parent=self)
+            if not el.get_attribute("class")
+        ]
+
+    def get_section_id_by_text(self, text):
+        for section in self.all_sections:
+            if section.text == text:
+                return section.value
+        raise NoSuchElementException("No section named {}".format(text))
+
+    def get_elements_in_section(self, text):
+        return self.browser.elements(
+            self.SECTION_ELEMENTS.format(self.get_section_id_by_text(text))
+        )
+
+    def select_item_in_section(self, section_text, item_text):
+        """ Selects an element in a section, only supports a single item
+
+        Args:
+            item_text: text of the item in the section
+            section_text: text of the section which contains the item
+        """
+        self.open()
+        # get the elements in the section
+        section_elements = self.get_elements_in_section(section_text.upper())
+        for el in section_elements:
+            try:
+                text_element = self.browser.element(".//span[contains(@class, 'text')]", parent=el)
+            except NoSuchElementException:
+                continue
+            if self.browser.text(text_element) == item_text:
+                self.browser.click(text_element)
+                self.close()
+                break
+        else:
+            raise SelectItemNotFound(
+                widget=self, item=item_text, options=[opt.text for opt in self.all_options]
+            )
+
+    def fill(self, values):
+        """ Values is assumed to have the tag category and the tag name """
+        before = self.selected_option
+        if len(values) == 3:
+            section_text = values[1]
+        else:
+            section_text = values[0]
+        item_text = values[-1]
+        self.select_item_in_section(section_text, item_text)
+        after = self.selected_option
+        return before != after
+
```

Comment:
Any read() implementation any different than BoostrapSelect?

Reply:
Nope, read is just 
```python,
    def read(self):
        if self.is_multiple:
            return self.all_selected_options
        else:
            return self.selected_option
```
`self.is_multiple` will be false for this widget, so it'll just return `self.selected_option`.